Adapt Python SDK to Typescript (E2B-30)

jzanecook opened the pull request:
> Thanks @jzanecook and congratulation for being our first contributor! Ran quick test and seems to be working. Good job!

Thanks!

> Is there a reason to prefer this over [OpenAPI generator](https://github.com/OpenAPITools/openapi-generator)? Did you try this [generator](https://openapi-generator.tech/docs/generators/nodejs-express-server/)?

No, I have not. Should I?

> @ValentaTomas correct me, if I'm wrong. The reason for static method is to have one global `Agent` wrapper and you can then conveniently import it from different files in the codebase without the need to pass it as an argument. At the moment it isn't that important, because there isn't any functionality why would you need importing the `Agent` yet.

I made the agent use static methods, but ran into issues when both were static, I referenced this article https://stackoverflow.com/a/64733963 and did it that way.

> * Let's name the package `agent-protocol`. I have the name reserved.

Done.

> * Can you add script for generating the boilerplate (models, routes, etc.)

I didn't use a script to generate it, I guess that's part of the OpenAPI generators?

> * What is the purpose of json-schema-to-typescript in Python SDK?

This was with the pydantic2ts package and should be removed.

> * When running `npm run fix` I am getting this errors
> 
> ```
> This rule requires the `strictNullChecks` compiler option to be turned on to function correctly  @typescript-eslint/strict-boolean-expressions
> ```

Fixed, I added this compiler option.

> 
> * You have `prettier` among scripts, but it isn't among dev dependencies

I had been using the installed prettier, not the package prettier, fixed now.

> * Remove `jzanecook-e2b-sdk-test` from dependecies

Whoopsies.

> * When trying to build I am getting this
> 
> ```
> Entry module "src/index.ts" is using named and default exports together. Consumers of your bundle will have to use `agentSdk["default"]` to access the default export, which may not be what you want. Use `output.exports: "named"` to disable this warning
> ```

This was a lot more effort than I was expecting lol, but at the end what I had to do was build CJS and ESM separately, reference the build script. I also ran into issues with the typings going into `dist/src` and erroring, so I added a little extra to the build script to make that go away.

> * It would be nice to have linted examples/ as well

Added the examples to the linting process in tsconfig, also removed them from the build.

jzanecook commented:
> I checked `microbundle` and it doesn't support default and named exports in one file. So I would prefer not mixing or using `tsup`. This looks too complicated and probably it's bypassing their design decision to not support it.

That's correct, it was from their github issues I found the resolution with the separated build. You're right though, it's too complicated. I wasn't able to get the `tsup` build to work at all, though.

jzanecook commented:
> > > Is there a reason to prefer this over [OpenAPI generator](https://github.com/OpenAPITools/openapi-generator)? Did you try this [generator](https://openapi-generator.tech/docs/generators/nodejs-express-server/)?
> > 
> > 
> > No, I have not. Should I?
> 
> I would prefer it. This way you would depend on the python package. If any change happens in the spec, you will have to change adjust python code and after that you can adjust js code.

Well that's a shame, I was going in the wrong direction the whole time...

> > I made the agent use static methods, but ran into issues when both were static, I referenced this article https://stackoverflow.com/a/64733963 and did it that way.
> 
> I added some code suggestions to copy the python behavior.

Code suggestions?

> Something like this [line](https://github.com/e2b-dev/sdk/blob/main/agent_client/python/scripts/generate.sh#L26). When we change the spec, you will just need to run the command to regenerate the code and maybe do some small adjustments. But it should make it much more maintainable.

Okay, I'll take a look and see what I can come up with.

jzanecook commented:
Can we separate these comments out into individual issues so that we can isolate them and clarify the requirements a little easier?

jakubno commented on agent/js/package.json:
```diff
@@ -1,24 +1,45 @@
-    "dev": "tsup --watch",
-    "fix": "npx eslint **/src/**/*.{ts,js} --fix",
+    "clean": "rm -rf ./dist",
+    "build": "rm -rf ./dist && microbundle --tsconfig tsconfig.json --no-sourcemap -f cjs,umd src/index.cjs.ts && microbundle --tsconfig tsconfig.json --no-sourcemap -f modern,esm src/index.ts && mv dist/src/* dist/ && rm -rf dist/{examples,src}",
```
```suggestion
    "build": "rm -rf ./dist && tsup",
```

with this config file works well for me 
```js
import { defineConfig } from 'tsup'

import * as packageJSON from './package.json'

const excludedPackages = []

export default defineConfig({
  entry: ['src/index.ts'],
  target: 'node16',
  platform: 'node',
  format: 'cjs'
})
```

jakubno reviewed (changes requested):
Sorry I didn't submit the suggestions 🙈

jakubno commented:
> Can we separate these comments out into individual issues so that we can isolate them and clarify the requirements a little easier?

Sure, let's get this done. I'll create an issue for generating the types and endpoints. This works and that's the most important thing! :tada:

jakubno commented:
So just address the suggestions, rebase on master and let's get it out 💪🏻

jzanecook commented:
Pretty sure I rebased properly here, I also:
- Made sure that the build with tsup (which I'm glad worked) also built the types so that if I ran the minimal example with `ts-node` it wouldn't yell at me.
- Removed `index.cjs.ts` since it's no longer needed.
- Added the proper typing to the static function in the Agent class so that eslint wouldn't yell at me.
- Successfully verified that nothing is yelling at me! :tada:

PS: It should be noted that eslint and prettier conflict. I've been pushing after prettier, but if you run eslint without prettier it will cause changes in all of the files. We may consider putting the eslint check and the prettier fix into a singular command, or tweaking the eslint to match prettier.

jakubno commented on agent/js/package.json:
```diff
@@ -1,24 +1,45 @@
+    "@typescript-eslint/eslint-plugin": "^5.62.0",
+    "eslint": "^8.45.0",
+    "eslint-config-standard-with-typescript": "^36.1.0",
+    "eslint-plugin-import": "^2.27.5",
+    "eslint-plugin-n": "^16.0.1",
+    "eslint-plugin-promise": "^6.1.1",
```
They aren't used

jakubno replied:
Add `eslint-config-prettier` to prevent the problem with `prettier`

jzanecook replied:
Done.

jakubno commented on agent/README.md:
```diff
@@ -20,3 +31,14 @@ pip install agent-protocol
+### Python
+[See the Python package](./python)
+
+### JavaScript/TypeScript
```
You can change this 😊

jzanecook replied:
I made it match the python section. :+1:

jakubno commented on agent/js/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+    project: "tsconfig.json",
+  },
+  rules: {
+    quotes: [2, "single", { avoidEscape: true }],
```
can you change it to single?
```suggestion
    quotes: [1, "single", { avoidEscape: true }],
```

jzanecook replied:
I actually went to double instead of single.

jakubno commented on agent/js/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+    es2021: true,
+  },
+  ignorePatterns: ["dist"],
+  extends: "standard-with-typescript",
```
To prevent collision between prettier and eslint
```suggestion
  extends: ["standard-with-typescript", "prettier"],
```

jzanecook replied:
Now they match, we are in unison.

jakubno commented:
Just last few nit picks and we'll merge it 🎉

jakubno commented:
Thanks @jzanecook for your contribution!
